### PR TITLE
Statistical weights in IndependentSource

### LINKED ISSUE #3194: Set statistical weights in IndependentSource
## Description
When multiple sources are used in the same simulation, the way to indicate the intensity of each one is currently through the parameter strength. This is used to calculate the probability of sampling each particle from each source with weight 1.
In cases where the difference in intensity is several orders of magnitude, it becomes almost impossible to obtain statistics from one of the sources. For this reason, it could be useful to allow setting the statistical weight of the source instead of the source's strength. This would result in equal sampling from each source, but each would be assigned the corresponding weight.

## Alternatives
Implement another weight management strategy or leave everything as it is. 

## Compatibility
There are no compatibility issues.

### PR DESCRIPTION
**Pull request by zoeprieto**
# Description

This is a solution to issue #3194. It adds the capability to set the value of the statistical weights for an `IndependentSource`. Especially useful when you have multiple sources with differences in several orders of magnitude of intensities. 

Fixes #3194

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks for this contribution @zoeprieto.  I'm not sure why `weight` is a distribution to be sampled and not just a constant like `strength`.   Your reasoning for this change as described in #3194 seems to suggest that it should be a constant for each `IndependentSource`

**Comment by zoeprieto** [comment]
Thank you for your response @gonuke! I had thought that for problems where you want to model a single source, it might be useful to make a weight distribution. But on second thought, that would be useful if I could simulate it coupled to another variable, such as energy. But as it is not implemented, the option should be only 1 value. I will work on that  :)
Thanks again.

**Review by gonuke — APPROVED** [review]
This looks good to me.  

I wonder about documenting the relationship between `strength` and `weight`. The most logical use would be for one of them to always be 1. That is, if you have multiple sources, the easiest to understand is all the sources have weight=1 and different strengths, or all the sources have strength=1 and different weights.

If they had different strengths as well as different weights, it will be a little tricky to interpret.

**Comment by zoeprieto** [comment]
Yes, I absolutely agree, they should not be used simultaneously. 
I updated the documentation. Let me know what you think.

**Comment by gonuke** [comment]
This discussion of documentation made me think of a different approach to this (sorry) - what if there was an option in the settings to just treat the strengths as weights.  Then a user could define the relative strengths just as they do now, but the sampling between the distributions would be handled differently just by changing that option.

(Note, I'm particularly interested in this topic right now because I have a student working on #3104 )

**Comment by zoeprieto** [comment]
I think what you are proposing is a better solution. Especially because there is no option to make a mistake by setting the strength variable simultaneously with weight. I have already implemented this as a first approximation. Let me know what you think.

File: openmc/source.py [hunk 1]
```diff
@@ -258,6 +258,8 @@ class IndependentSource(SourceBase):
         Energy distribution of source sites
     time : openmc.stats.Univariate
         time distribution of source sites
+    weight : float
+        weight value of source sites
```
Comment (gonuke) [inline openmc/source.py:262]: I don't think we use this anymore....

File: openmc/source.py [hunk 2]
```diff
@@ -292,6 +294,8 @@ class IndependentSource(SourceBase):
         Energy distribution of source sites
     time : openmc.stats.Univariate or None
         time distribution of source sites
+    weight : float or None
+        weight value of source sites
```
Comment (gonuke) [inline openmc/source.py:298]: Is this still used?

**Review by gonuke — CHANGES_REQUESTED** [review]
This looks a lot simpler.  Interested in various opinions on the best name for this.  I think `strengths_as_weights` makes a little more sense, but there may be a lot of different thoughts?

File: openmc/source.py (hunk 1 above)
Comment (zoeprieto) [reply openmc/source.py:262]: You're right, thanks!

File: openmc/source.py (hunk 2 above)
Comment (zoeprieto) [reply openmc/source.py:298]: No, thank you for checking. It's already deleted.

File: src/source.cpp [hunk 3]
```diff
@@ -611,17 +615,27 @@ SourceSite sample_external_source(uint64_t* seed)
...
-        break;
+      if (openmc::settings::strength_to_weights) {
+        c += 1.0;
+        if (xi < c)
+          break;
+      } else {
+        c += model::external_sources[i]->strength();
+        if (xi < c)
+          break;
+      }
```
[the remark this answers is not in the document]
Comment (zoeprieto) [reply src/source.cpp:638]: changed, thanks :)

**Comment by zoeprieto** [comment]
Thanks for the comments, I have renamed the variable to `strengths_as_weights` and fixed the changes you pointed out.

**Review by gonuke — APPROVED** [review]
This looks good to me - it will be up to @pshriwise or @paulromano to give it a final review/approval.

**Comment by paulromano** [comment]
Thanks @gonuke. I'll put this on my list and should be able to get to it this week.

**Review by paulromano — APPROVED** [review]
Thanks for this PR @zoeprieto! I made a few changes/fixes/improvements here as follows:

- Rather than this being something specific to `IndependentSource`, I've raised the logic for applying the source strength into the `sample_external_source` function so that it works with any source class
- There is a piece of code in tally.cpp where we apply the source strength to tallies as a normalization after the run is done. Since the strength is applied during transport for this option, I added a conditional to ensure that the strength is not applied twice (which would result in wrong results). I also added a new test to check for this.
- I changed the name to `uniform_source_sampling`, which I think more clearly expresses to the user what it is about.
- I used your `ParticleList` class in the tests to simplify them :smile: 

If you and @gonuke are satisfied with these changes, this should be good to merge.

**Comment by zoeprieto** [comment]
I think the changes are great @paulromano . Thanks for generalizing it to all type of sources and sorry about the normalization of the tallies, I missed it.

**Comment by gonuke** [comment]
I'm on the fence about `uniform_source_sampling` because it doesn't convey anything about bias or weight adjustment.  I worry a user may interpret this as uniform strength and not a biased uniform sampling...  not sure about a better name, though.

**Comment by paulromano** [comment]
> a user may interpret this as uniform strength and not a biased uniform sampling

I put "uniform" and "sampling" in the name so as to avoid that interpretation. The docstring in `Settings` is also pretty explicit about what it does, which should also help avoid misinterpretation, although if you think it could be clearer I'm open to suggestions.

**Comment by paulromano** [comment]
@gonuke Are you OK with merging this as is?

**Comment by gonuke** [comment]
I'm OK with this as is

**Pull request merged**